core/shared/src/main/scala/scalaz/zio/ExitResult.scala

```diff
@@ -9,56 +10,192 @@ package scalaz.zio
+
+  sealed abstract class Cause[+E] extends Product with Serializable { self =>
+    import Cause._
+    final def ++[E1 >: E](that: Cause[E1]): Cause[E1] =
+      Then(self, that)
+
+    final def &&[E1 >: E](that: Cause[E1]): Cause[E1] =
+      Both(self, that)
+
+    final def map[E1](f: E => E1): Cause[E1] = self match {
+      case Failure(value)      => Failure(f(value))
+      case c @ Exception(_)    => c
+      case c @ Interruption(_) => c
 
-  final def zipWith[E1 >: E, B, C](that: ExitResult[E1, B])(f: (A, B) => C): ExitResult[E1, C] = (self, that) match {
-    case (ExitResult.Completed(a), ExitResult.Completed(b))   => ExitResult.Completed(f(a, b))
-    case (ExitResult.Failed(e, ts), rb)                       => ExitResult.Failed(e, combine(ts, rb))
-    case (ExitResult.Terminated(ts), rb)                      => ExitResult.Terminated(combine(ts, rb))
-    case (ExitResult.Completed(_), ExitResult.Failed(e, ts))  => ExitResult.Failed(e, ts)
-    case (ExitResult.Completed(_), ExitResult.Terminated(ts)) => ExitResult.Terminated(ts)
+      case Then(left, right) => Then(left.map(f), right.map(f))
+      case Both(left, right) => Both(left.map(f), right.map(f))
+    }
+
+    final def failures[E1 >: E]: Set[E1] =
+      self.fold(Set.empty[E1]) {
+        case (z, Failure(v)) => z + v
+      }
+
+    final def exceptions: List[Throwable] =
+      self
+        .fold(List.empty[Throwable]) {
+          case (z, Exception(v)) => v :: z
+        }
+        .reverse
+
+    final def interruptions: Option[List[Throwable]] =
+      self
+        .fold(Option.empty[List[Throwable]]) {
+          case (z, Interruption(Some(v))) =>
+            z.fold(Some(v :: Nil))(vs => Some(v :: vs))
+          case (z, Interruption(None)) =>
+            z.fold(Some(List[Throwable]()))(vs => Some(vs))
+        }
+        .map(_.reverse)
+
+    final def fold[Z](z: Z)(f: PartialFunction[(Z, Cause[E]), Z]): Z =
+      (f.lift(z -> self).getOrElse(z), self) match {
+        case (z, Then(left, right)) => right.fold(left.fold(z)(f))(f)
+        case (z, Both(left, right)) => right.fold(left.fold(z)(f))(f)
+
+        case (z, _) => z
+      }
+
+    final def toThrowable(causes: List[Throwable] = Nil): Throwable =
+      self match {
+        case Failure(error)      => Errors.UnhandledError(error, causes)
+        case Exception(defect)   => Errors.TerminatedFiber(defect, causes)
+        case Interruption(cause) => Errors.InterruptedFiber(cause.map(_ :: Nil).getOrElse(Nil), causes)
+        case Then(left, right)   => left.toThrowable(right.toThrowable() :: causes)
+        case Both(left, right)   => Errors.ParallelFiberError(left.toThrowable(), right.toThrowable(), causes)
+      }
+
+    final def failure: Option[E] = self match {
+      case Failure(e)      => Some(e)
+      case Exception(_)    => None
+      case Interruption(_) => None
+      case Then(left, _)   => left.failure
+      case Both(_, _)      => None
```

Comment:
A wildcard pattern for all but the two special cases would do here. Also worth adding a `@tailrec` annotation.